lib/app/dispatcher.rb

```diff
@@ -0,0 +1,83 @@
+#
+# Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+# or more contributor license agreements. Licensed under the Elastic License;
+# you may not use this file except in compliance with the Elastic License.
+#
+
+# frozen_string_literal: true
+
+require 'active_support'
+require 'concurrent-ruby'
+require 'connectors'
+require 'core'
+require 'utility'
+require 'app/config'
+
+module App
+  class Dispatcher
+    POLL_IDLING = (App::Config[:idle_timeout] || 60).to_i
+    TERMINATION_TIMEOUT = (App::Config[:termination_timeout] || 60).to_i
+
+    attr_reader :scheduler
+
+    def initialize
+      @pool = Concurrent::ThreadPoolExecutor.new(
+        min_threads: 5,
+        max_threads: 10,
+        max_queue: 100,
+        fallback_policy: :abort
+      )
+      @is_shutting_down = false
+      @heartbeats = {}
+    end
+
+    def start!
+      # dispatcher-level checks
+      Core::ElasticConnectorActions.ensure_connectors_index_exists
+      Core::ElasticConnectorActions.ensure_job_index_exists
+
+      @scheduler ||= Core::NativeScheduler.new(POLL_IDLING)
+      @scheduler.when_triggered do |connector_settings|
+        service_type = connector_settings.service_type
+        connector_id = connector_settings.id
+
+        # connector-level checks
+        Core::ElasticConnectorActions.ensure_content_index_exists(connector_settings.index_name)
+        raise "[#{service_type}] is not a supported connector" unless Connectors::REGISTRY.registered?(service_type)
+
+        @pool.post do
+          send_heartbeat(connector_id, service_type)
```

Comment:
heartbeat should be sent every half hour according to the [protocol](https://github.com/elastic/connectors-ruby/blob/main/docs/CONNECTOR_PROTOCOL.md), but I'm if we tackle it in a separate PR.